### PR TITLE
.Net: MS AI Azure Inference Connector Update

### PR DESCRIPTION
**Pull request by RogerBarreto**
#### Updates Azure AI Inference Connector to use `Microsoft Extensions AI`.

- Enables Logging 
- Enables Function Calling
- Adds extra extension for `AzureAIInferenceChatClient`
- Updates Demos/Concepts with the new pattern.

#### ⚠️ Breaking Changes
- ChatCompletion Service is Deprecated (Obsoleted) in favor of 
    - `Microsoft.Extensions.AI.AzureAIInferenceChatClient` `.AsChatCompletionService()` or 
    - `Azure.AI.Inference.ChatCompletionsClient` `.AsChatClient().AsChatCompletionService()`
    
- `modelId` is required for all ChatCompletion extensions

### CODE REVIEW

File: dotnet/src/Connectors/Connectors.AzureAIInference/Extensions/AzureAIInferenceServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -18,71 +19,107 @@ namespace Microsoft.SemanticKernel;
...
-        AzureAIInferenceChatCompletionService Factory(IServiceProvider serviceProvider, object? _) =>
-            new(modelId,
-                apiKey,
-                endpoint,
-                HttpClientProvider.GetHttpClient(httpClient, serviceProvider),
-                serviceProvider.GetService<ILoggerFactory>());
+        services.AddKeyedSingleton<IChatCompletionService>(serviceId, (serviceProvider, _) =>
+        {
+            var chatClientBuilder = new ChatClientBuilder()
+               .UseFunctionInvocation(config =>
```
Comment (markwallace-microsoft) [inline dotnet/src/Connectors/Connectors.AzureAIInference/Extensions/AzureAIInferenceServiceCollectionExtensions.cs:44]: We may want to switch in our own function calling middleware which supports our function calling abstractions and filters. Let's discuss this post Ignite.

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/samples/Concepts/ChatCompletion/AzureAIInference_ChatCompletion.cs [hunk 2]
```diff
@@ -15,9 +16,13 @@ public async Task ServicePromptAsync()
...
         Console.WriteLine("======== Azure AI Inference - Chat Completion ========");
 
-        var chatService = new AzureAIInferenceChatCompletionService(
-            endpoint: new Uri(TestConfiguration.AzureAIInference.Endpoint),
-            apiKey: TestConfiguration.AzureAIInference.ApiKey);
+        Assert.NotNull(TestConfiguration.AzureAIInference.ApiKey);
+
+        var chatService = new ChatCompletionsClient(
+                endpoint: new Uri(TestConfiguration.AzureAIInference.Endpoint),
+                credential: new Azure.AzureKeyCredential(TestConfiguration.AzureAIInference.ApiKey))
```
Comment (SergeyMenshykh) [inline dotnet/samples/Concepts/ChatCompletion/AzureAIInference_ChatCompletion.cs:23]: Should we use api key instead of credential for samples? Same below.

**Review by SergeyMenshykh — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.AzureAIInference/Services/AzureAIInferenceChatCompletionService.cs [hunk 3]
```diff
@@ -16,9 +17,11 @@ namespace Microsoft.SemanticKernel.Connectors.AzureAIInference;
 /// <summary>
 /// Chat completion service for Azure AI Inference.
 /// </summary>
+[Obsolete("Dedicated AzureAIInferenceChatCompletionService is deprecated. Use OllamaApiClient.AsChatCompletionService() instead.")]
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AzureAIInference/Services/AzureAIInferenceChatCompletionService.cs:20]: Should it be like this?
```suggestion
[Obsolete("Dedicated AzureAIInferenceChatCompletionService is deprecated. Use ChatCompletionsClient.AsChatClient().AsChatCompletionService() instead.")]
```

File: dotnet/src/Connectors/Connectors.AzureAIInference/Services/AzureAIInferenceChatCompletionService.cs [hunk 4]
```diff
@@ -77,20 +108,40 @@ public AzureAIInferenceChatCompletionService(
...
+        => this._chatService.GetChatMessageContentsAsync(chatHistory, executionSettings, kernel, cancellationToken);
 
     /// <inheritdoc/>
     public IAsyncEnumerable<StreamingChatMessageContent> GetStreamingChatMessageContentsAsync(ChatHistory chatHistory, PromptExecutionSettings? executionSettings = null, Kernel? kernel = null, CancellationToken cancellationToken = default)
-        => this._core.GetStreamingChatMessageContentsAsync(chatHistory, executionSettings, kernel, cancellationToken);
+        => this._chatService.GetStreamingChatMessageContentsAsync(chatHistory, executionSettings, kernel, cancellationToken);
+
+    #region Private
+
+    private const int MaxInflightAutoInvokes = 128;
```
Comment (dmytrostruk) [inline dotnet/src/Connectors/Connectors.AzureAIInference/Services/AzureAIInferenceChatCompletionService.cs:144]: This constant exists in both `AzureAIInferenceChatCompletionService.cs` and `AzureAIInferenceServiceCollectionExtensions.cs`. Should we define a single constant in separate file and re-use it in multiple places?

File: dotnet/src/IntegrationTests/Connectors/AzureAIInference/AzureAIInferenceChatCompletion_FunctionCallingTests.cs [hunk 5]
```diff
@@ -0,0 +1,362 @@
...
+
+        // Act
+        var result = await kernel.InvokePromptAsync("What is the current temperature in Dublin, Ireland, in Fahrenheit?", new(settings));
+
+        // Assert
+        Assert.NotNull(result);
+        Assert.Contains("42.8", result.GetValue<string>(), StringComparison.InvariantCulture); // The WeatherPlugin always returns 42.8 for Dublin, Ireland.
+    }
+
+    [Fact]//(Skip = "For manual verification only")]
```
Comment (dmytrostruk) [inline dotnet/src/IntegrationTests/Connectors/AzureAIInference/AzureAIInferenceChatCompletion_FunctionCallingTests.cs:47]: Double checking if these tests should be enabled by default. If yes, maybe it's worth to remove the comment. Same in other places.

**Review by dmytrostruk — APPROVED** [review]

File: dotnet/samples/Concepts/ChatCompletion/AzureAIInference_ChatCompletion.cs (hunk 2 above)
Comment (RogerBarreto) [reply dotnet/samples/Concepts/ChatCompletion/AzureAIInference_ChatCompletion.cs:23]: There's no option for passing the `apiKey`, it needs to be a class, in this case an `AzureKeyCredential`.

File: dotnet/src/IntegrationTests/Connectors/AzureAIInference/AzureAIInferenceChatCompletion_FunctionCallingTests.cs (hunk 5 above)
Comment (RogerBarreto) [reply dotnet/src/IntegrationTests/Connectors/AzureAIInference/AzureAIInferenceChatCompletion_FunctionCallingTests.cs]: Good catch, added Skip, as currently I wasn't able to deploy any model in Azure AI Studio capable of function calling, including llama3.2 with a weird error message:

```http
HTTP/1.1 400 Bad Request
Content-Length: 246
Content-Type: application/json
x-ms-rai-invoked: false
x-request-id: b0bf8687-9587-4cc6-8cd8-66d2bf86e22d
ms-azureml-model-error-reason: model_error
ms-azureml-model-error-statuscode: 400
x-ms-client-request-id: 4796effe-4005-40fd-aaf1-d5ddd19159fd
Request-Context: appId=cid-v1:38659f32-9fc9-4714-92e6-1412a97fbbfc
azureml-model-session: dep-90b-vision-240926224530
azureml-model-group: llama-3-2-90b-scaleset
Date: Mon, 11 Nov 2024 15:39:51 GMT
Connection: close

{  
"error": {
    "code": "Bad Request",
    "message": "{\"object\":\"error\",\"message\":\"\\\"auto\\\" tool choice requires --enable-auto-tool-choice and --tool-call-parser to be set\",\"type\":\"BadRequestError\",\"param\":null,\"code\":400}",
    "status": 400
  }
}
```

**Review by markwallace-microsoft — APPROVED** [review]

**Review by markwallace-microsoft — APPROVED** [review]

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**